# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: toolkit/xre/nsEmbedFunctions.cpp
# Commit: 2cf5cad90099
# Full Hash: 2cf5cad900992236d596bfb329099742ff15f689
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2024-06-20 04:08:16
# Regressor Bug: 1728331
# File Overlap Count: 1
# Description:
#   Bug 1728331 - Part 4: Make ContentParent KeepAlives explicit with RAII references, r=smaug,dom-worker-reviewers,asuth
#   
#   This is a fairly significant patch, however it would be difficult to break it
#   down into smaller patches:
#   
# ==============================================================================

diff -r d920c2d72d00 -r 2cf5cad90099 toolkit/xre/nsEmbedFunctions.cpp
--- a/toolkit/xre/nsEmbedFunctions.cpp	Wed Jun 19 20:14:49 2024 +0000
+++ b/toolkit/xre/nsEmbedFunctions.cpp	Wed Jun 19 20:14:50 2024 +0000
@@ -63,6 +63,7 @@
 #include "mozilla/AbstractThread.h"
 #include "mozilla/FilePreferences.h"
 #include "mozilla/IOInterposer.h"
+#include "mozilla/NeverDestroyed.h"
 #include "mozilla/ProcessType.h"
 #include "mozilla/RDDProcessImpl.h"
 #include "mozilla/ipc/UtilityProcessImpl.h"
@@ -140,6 +141,7 @@
 
 using mozilla::dom::ContentParent;
 using mozilla::dom::ContentProcess;
+using mozilla::dom::UniqueContentParentKeepAlive;
 
 using mozilla::gmp::GMPProcessChild;
 
@@ -739,22 +741,25 @@
 }
 
 namespace {
-ContentParent* gContentParent;  // long-lived, manually refcounted
+UniqueContentParentKeepAlive& TestShellContentParent() {
+  static NeverDestroyed<UniqueContentParentKeepAlive> sContentParent;
+  return *sContentParent;
+}
+
 TestShellParent* GetOrCreateTestShellParent() {
-  if (!gContentParent) {
+  if (!TestShellContentParent()) {
     // Use a "web" child process by default.  File a bug if you don't like
     // this and you're sure you wouldn't be better off writing a "browser"
     // chrome mochitest where you can have multiple types of content
     // processes.
-    RefPtr<ContentParent> parent =
+    TestShellContentParent() =
         ContentParent::GetNewOrUsedBrowserProcess(DEFAULT_REMOTE_TYPE);
-    parent.forget(&gContentParent);
-  } else if (gContentParent->IsShuttingDown()) {
+  } else if (TestShellContentParent()->IsShuttingDown()) {
     return nullptr;
   }
-  TestShellParent* tsp = gContentParent->GetTestShellSingleton();
+  TestShellParent* tsp = TestShellContentParent()->GetTestShellSingleton();
   if (!tsp) {
-    tsp = gContentParent->CreateTestShell();
+    tsp = TestShellContentParent()->CreateTestShell();
   }
   return tsp;
 }
@@ -784,15 +789,15 @@
 }
 
 bool XRE_ShutdownTestShell() {
-  if (!gContentParent) {
+  if (!TestShellContentParent()) {
     return true;
   }
   bool ret = true;
-  if (gContentParent->IsAlive()) {
-    ret = gContentParent->DestroyTestShell(
-        gContentParent->GetTestShellSingleton());
+  if (TestShellContentParent()->IsAlive()) {
+    ret = TestShellContentParent()->DestroyTestShell(
+        TestShellContentParent()->GetTestShellSingleton());
   }
-  NS_RELEASE(gContentParent);
+  TestShellContentParent().reset();
   return ret;
 }
 
